codegen/cmd/injection-gen/generators/reconciler_reconciler.go

```diff
@@ -410,8 +411,15 @@ func (r *reconcilerImpl) Reconcile(ctx {{.contextContext|raw}}, key string) erro
 	original, err := getter.Get(s.name)
 
 	if {{.apierrsIsNotFound|raw}}(err) {
-		// The resource may no longer exist, in which case we stop processing.
+		// The resource may no longer exist, in which case we stop processing and call
+		// the ObserveDeletion handler if appropriate.
 		logger.Debugf("Resource %q no longer exists", key)
+		if del, ok := r.reconciler.({{.reconcilerOnDeletionInterface|raw}}); ok {
+			return del.ObserveDeletion(ctx, {{.typesNamespacedName|raw}}{
```

Comment:
I would not return anything here, no reason to requeue on error, the resource is gone

Reply:
Doesn't that make retrying errors a bit easier? It'd go through the same codepath again and error handling would be similar to the usual reconciler functions.

Reply:
the observer does not need to retry, I see this as an FYI and it should have no fail path, and should not be able to requeue the key.

Do you have a usecase for an observer that is able to re-enqueue?

Reply:
Yep, https://github.com/knative-sandbox/net-kourier/blob/7387060469bcbeccb8e99e8abe141a2647c507c3/pkg/reconciler/ingress/ingress.go#L115-L150 for instance. Updating the envoy config there **can** error out and I'm not quite clear how I'd resolve such error if the reconciler framework doesn't do it for me.

I do see your point that this can be handled as more of an FYI, but I do believe it's valuable to offer the error return + retry facilities that reconcilers usually have as well.

Reply:
my pushback is because it does not follow the other readonly methods we define: 

```
type ReadOnlyInterface interface {
	// ObserveKind implements logic to observe {{.type|raw}}.
	// This method should not write to the API.
	ObserveKind(ctx {{.contextContext|raw}}, o *{{.type|raw}}) {{.reconcilerEvent|raw}}
}
// ReadOnlyFinalizer defines the strongly typed interfaces to be implemented by a
// controller finalizing {{.type|raw}} if they want to process tombstoned resources
// even when they are not the leader.  Due to the nature of how finalizers are handled
// there are no guarantees that this will be called.
type ReadOnlyFinalizer interface {
	// ObserveFinalizeKind implements custom logic to observe the final state of {{.type|raw}}.
	// This method should not write to the API.
	ObserveFinalizeKind(ctx {{.contextContext|raw}}, o *{{.type|raw}}) {{.reconcilerEvent|raw}}
}
```

Reply:
interesting, ok so the diff here is we are talking about Kind based reconcilers, but yours is for queue key based... until this point we have not been leaking the queue key. But I suppose you no longer have the deleted object to make a:

```
type ReadOnlyDeletion interface {
	ObserveDeleteKind(ctx {{.contextContext|raw}}, o *{{.type|raw}}) {{.reconcilerEvent|raw}}
}
```

Reply:
Right, the deleted object is already gone.

Reply:
In my opinion, we should not allow an observation method to return an error that will cause the key to be requeued.
Additionally, I do not think we should pass the key as a string to a consumer because it leaks more information that I would like for future changes that are "planned".

How about this spin on this idea:

```
ObserveDeletion(ctx context.Context, ref duckv1.KReference)
```

changes would be:

1. Drop the return error.
2. Drop the queue key string in favor of a object ref. (reason being: the downstream reconciler should need to know how we parse keys, leaving us free to change the queue key scheme in the future)

Reply:
FWIW, we're not passing strings, we pass a NamespacedName, which is already a parsed version of the string. If we don't return errors (and thus disallow requeuing) how are observers supposed to handle errors? To take the example from above: If we fail to update the envoy config, I'd actually like to retry that. I can't retry in the handler though, as that'd block a reconciler thread.

Reply:
I need a second opinion. @vaikas

Reply:
I am just an idiot... sorry.
/lgtm
/approve

Reply:
Just to make sure I understand. Is it correct to say that the problem we're trying to solve here is:
 1. We have a resource (call it "original" resource) that creates another resource. I reckon this resource is maybe not a k8s resource, because for managing ownership, we'd use the OwnerRef?
 2. That resource then needs to be cleaned up when the "original" resource gets deleted. Or Something needs to be updated that's not a k8s resource?
 3. If something goes wrong during that delete phase, things are left in an inconsistent state?
 
Is that the problem here?
IFF that's the problem my probably dump question is why not use finalizers here then to ensure that the object will get reconciled properly.

Reply:
@vaikas see the linked issue for the motivation of this. This tries to solve the problem of cleaning up in-memory state consistent with the deletion of K8s objects. Finalizers are not guaranteed to be observed by all replicas and directly reacting to the informer events isn't guaranteed to not race an update event.